framework/include/interfaces/FunctorInterface.h

```diff
@@ -0,0 +1,311 @@
+   * @return The requested element type data indexed at the requested quadrature point. As for the
+   * \p qp overload, there is a caveat: any variables involved in the functor evaluation must have
+   * their requested element data type properly pre-initialized at the desired quadrature point
+   */
+  virtual T evaluate(const std::tuple<Moose::ElementType, unsigned int, SubdomainID> & tqp,
+                     unsigned int state) const = 0;
+
+private:
+  /**
+   * clear cache data
+   */
+  void clearCacheData();
+
+  /// How often to clear the material property cache
+  std::set<ExecFlagType> _clearance_schedule;
+
+  // Data for traditional element-quadrature point property evaluations which are useful for caching
+  // implementation
+
+  /// Current quadrature point element (current cache key)
+  mutable dof_id_type _current_qp_elem = DofObject::invalid_id;
+
+  /// Current quadrature point element data (current cache value)
+  mutable std::vector<std::pair<bool, T>> * _current_qp_elem_data = nullptr;
+
+  /// Cached element quadrature point functor property evaluations. The map key is the element
+  /// id. The map values should have size corresponding to the number of quadrature points on the
+  /// element. The vector elements are pairs. The first member of the pair indicates whether a
+  /// cached value has been computed. The second member of the pair is the (cached) value. If the
+  /// boolean is false, then the value cannot be trusted
+  mutable std::unordered_map<dof_id_type, std::vector<std::pair<bool, T>>> _qp_to_value;
+};
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const Elem * const & elem, const unsigned int state) const
+{
+  return evaluate(elem, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const ElemAndFaceArg & elem_and_face,
+                                const unsigned int state) const
+{
+  return evaluate(elem_and_face, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const FaceArg & face, const unsigned int state) const
+{
+  return evaluate(face, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const QpArg & elem_and_qp, const unsigned int state) const
+{
+  if (_clearance_schedule.count(EXEC_ALWAYS))
+    return evaluate(elem_and_qp, state);
+
+  const auto elem_id = std::get<0>(elem_and_qp)->id();
+  if (elem_id != _current_qp_elem)
+  {
+    _current_qp_elem = elem_id;
+    _current_qp_elem_data = &_qp_to_value[elem_id];
+  }
+  auto & qp_elem_data_ref = *_current_qp_elem_data;
+  const auto qp = std::get<1>(elem_and_qp);
+
+  // Check and see whether we even have sized for this quadrature point. If we haven't then we must
+  // evaluate
+  if (qp >= qp_elem_data_ref.size())
+  {
+    mooseAssert(qp == qp_elem_data_ref.size(),
+                "I believe that we always iterate over quadrature points in a contiguous fashion");
+    qp_elem_data_ref.resize(qp + 1);
```

Comment:
This could cause a lot of memory re-allocation and copying - of potentially large material properties.  Is there any way to at least `reserve` the max number of qps in the problem (something we have on FEProblem).

Reply:
The allocation speed cost would only be paid in the beginning of the simulation because I do not clear the data structures when we clear the caches; instead the evaluated boolean is just toggled to `false`

Reply:
But I could do `resize(qrule->n_points())` for sure to make sure we cut the number of calls to `resize` by `n_points()`

Reply:
Recording this for comparison. Here is with one-off resizing
```
(pprof) list queryQpCache
Total: 56.14s
ROUTINE ======================== FunctorInterface::queryQpCache in /home/lindad/projects/moose/framework/build/header_symlinks/FunctorInterface.h
     330ms      3.70s (flat, cum)  6.59% of Total
         .          .    272:                                  const SpaceArg & space,
         .          .    273:                                  const TimeArg & time) const
         .          .    274:{
         .          .    275:  // Check and see whether we even have sized for this quadrature point. If we haven't then we must
         .          .    276:  // evaluate
      10ms       10ms    277:  if (qp >= qp_cache_data.size())
         .          .    278:  {
         .          .    279:    mooseAssert(qp_cache_data.size() == qp,
         .          .    280:                "We should only be resizing by one because we iterate over quadrature points in a "
         .          .    281:                "contiugous fashion");
         .       30ms    282:    qp_cache_data.resize(qp + 1);
         .          .    283:    auto & pr = qp_cache_data.back();
         .      500ms    284:    pr.second = evaluate(space, time);
         .          .    285:    pr.first = true;
         .          .    286:    return pr.second;
         .          .    287:  }
         .          .    288:
         .          .    289:  // We've already sized for this qp, so let's see whether we have a valid cache value
         .          .    290:  auto & pr = qp_cache_data[qp];
     300ms      300ms    291:  if (pr.first)
      20ms      110ms    292:    return pr.second;
         .          .    293:
         .          .    294:  // No valid cache value so evaluate
         .      2.39s    295:  pr.second = evaluate(space, time);
         .          .    296:  pr.first = true;
         .      360ms    297:  return pr.second;
         .          .    298:}
```

Reply:
Cut off .14 seconds!
```
(pprof) list queryQpCache
Total: 55.70s
ROUTINE ======================== FunctorInterface::queryQpCache in /home/lindad/projects/moose/framework/build/header_symlinks/FunctorInterface.h
     400ms      3.56s (flat, cum)  6.39% of Total
         .          .    267:}
         .          .    268:
         .          .    269:template <typename T>
         .          .    270:template <typename SpaceArg, typename TimeArg>
         .          .    271:T
      80ms       80ms    272:FunctorInterface<T>::queryQpCache(const unsigned int qp,
         .          .    273:                                  const QBase & qrule,
         .          .    274:                                  std::vector<std::pair<bool, T>> & qp_cache_data,
         .          .    275:                                  const SpaceArg & space,
         .          .    276:                                  const TimeArg & time) const
         .          .    277:{
         .          .    278:  // Check and see whether we even have sized for this quadrature point. If we haven't then we must
         .          .    279:  // evaluate
         .          .    280:  if (qp >= qp_cache_data.size())
         .          .    281:  {
         .       10ms    282:    qp_cache_data.resize(qrule.n_points(), std::make_pair(false, T()));
         .          .    283:    auto & pr = qp_cache_data[qp];
         .      470ms    284:    pr.second = evaluate(space, time);
         .          .    285:    pr.first = true;
         .          .    286:    return pr.second;
         .          .    287:  }
         .          .    288:
         .          .    289:  // We've already sized for this qp, so let's see whether we have a valid cache value
         .       10ms    290:  auto & pr = qp_cache_data[qp];
     230ms      230ms    291:  if (pr.first)
         .          .    292:    return pr.second;
         .          .    293:
         .          .    294:  // No valid cache value so evaluate
         .      2.26s    295:  pr.second = evaluate(space, time);
         .          .    296:  pr.first = true;
         .      410ms    297:  return pr.second;
      90ms       90ms    298:}
```